enterprise/backend/src/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions.clj

```diff
@@ -259,10 +260,24 @@
                      (query-perms/required-perms-for-query (:dataset-query (lib.metadata.protocols/card (qp.store/metadata-provider) card-id))
                                                  :throw-exceptions? true))
 
-    (let [table-ids (sandbox->table-ids sandbox)]
-      {:perms/view-data (zipmap table-ids (repeat :unrestricted))
+    (let [table-ids (sandbox->table-ids sandbox)
+          table-id->db-id (if (seq table-ids)
+                            (m/index-by :id (t2/select :model/Table :id [:in table-ids]))
```

Comment:
The DB call here makes me a little nervous since we've had recent perf issues in sandboxing code. Two ideas:
* Use `metabase.models.database/table-id->database-id` which is memoized per app DB. Issue there is that if the cache isn't populated it'll do a separate DB call for each table. Maybe we could add a batched version that does a single DB call for anything not in the cache but that sounds more complex.
* Update `metabase-enterprise.sandbox.api.util/enforced-sandboxes-for-user` to fetch DB IDs via a join when populating the sandbox cache for a given request. I like this idea better.

Reply:
I am working on implementing the second option here. I noticed since we already hydrate tables in the query for enforced-sandboxes-for-users, there's no need to join (we already get the table-id <-> db-id info. 

But! I noticed something interesting: I have sandboxed orders (see screenshot), and when I go to save a vanilla query on orders using a nomral user, we call `sandbox->required-perms`, where the sandbox and `@data-perms/*sandboxes-for-user*` have different sets of table ids.

```clojure
;; sandbox has:
{3 1, 5 1}

;; @data-perms/*sandboxes-for-user* has:
(map
 (comp (juxt :id :db_id) :table)
 (@data-perms/*sandboxes-for-user*)) => ([5 1])
```


![Screenshot](https://github.com/user-attachments/assets/2cb4b18e-4e6a-4a5f-ad39-cdb4321150cf)



This doesn't seem to find the People table:
``` 
;; sandbox has:
{3 1, 5 1}

;; @data-perms/*sandboxes-for-user* has
(map
 (comp (juxt :id :db_id) :table)
 (@data-perms/*sandboxes-for-user*)) => ([5 1])
```

Since this seems to not be a 1:1 drop in, I will start writing option 1. Curious to hear your thoughts